<center>
<img src="../../img/ods_stickers.jpg" />
    
## [mlcourse.ai](https://mlcourse.ai) – Open Machine Learning Course 

Author: [Yury Kashnitskiy](https://yorko.github.io). This material is subject to the terms and conditions of the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license. Free use is permitted for any non-commercial purpose.

# <center> Topic 1. Exploratory data analysis with Pandas
## <center>Practice. Analyzing "Titanic" passengers

**Fill in the missing code ("You code here") and choose answers in a [web-form](https://docs.google.com/forms/d/16EfhpDGPrREry0gfDQdRPjoiQX9IumaL2mPR0rcj19k/edit).**

In [6]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
pd.set_option("display.precision", 2)

**Read data into a Pandas DataFrame**

In [8]:
data = pd.read_csv('../practical/titanic_train.csv',
                  index_col='PassengerId')
print('Ready')

Ready


**First 5 rows**

In [9]:
data.head(5)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket   Fare Cabin Embarked  
PassengerId                                                        
1                1      0         A/5 21171   7.25   NaN        S  
2                1      0          PC 17599  71.28   C85        C  
3                0      0  STON/O2. 3101282   7.92   NaN        S  
4                1      0            113803  53.10  C123        S  
5                0      0            373450   8.05   NaN        S

In [10]:
data.describe()

Survived  Pclass     Age   SibSp   Parch    Fare
count    891.00  891.00  714.00  891.00  891.00  891.00
mean       0.38    2.31   29.70    0.52    0.38   32.20
std        0.49    0.84   14.53    1.10    0.81   49.69
min        0.00    1.00    0.42    0.00    0.00    0.00
25%        0.00    2.00   20.12    0.00    0.00    7.91
50%        0.00    3.00   28.00    0.00    0.00   14.45
75%        1.00    3.00   38.00    1.00    0.00   31.00
max        1.00    3.00   80.00    8.00    6.00  512.33

**Let's select those passengers who embarked in Cherbourg (Embarked=C) and paid > 200 pounds for their ticker (fare > 200).**

Make sure you understand how actually this construction works.

In [9]:
data[(data['Embarked'] == 'S') & (data.Fare < 200)].head()

Survived  Pclass                                          Name  \
PassengerId                                                                   
1                   0       3                       Braund, Mr. Owen Harris   
3                   1       3                        Heikkinen, Miss. Laina   
4                   1       1  Futrelle, Mrs. Jacques Heath (Lily May Peel)   
5                   0       3                      Allen, Mr. William Henry   
7                   0       1                       McCarthy, Mr. Timothy J   

                Sex   Age  SibSp  Parch            Ticket   Fare Cabin  \
PassengerId                                                              
1              male  22.0      1      0         A/5 21171   7.25   NaN   
3            female  26.0      0      0  STON/O2. 3101282   7.92   NaN   
4            female  35.0      1      0            113803  53.10  C123   
5              male  35.0      0      0            373450   8.05   NaN   
7              male  54.0      0      0             17463  51.86   E46   

            Embarked  
PassengerId           
1                  S  
3                  S  
4                  S  
5                  S  
7                  S

**We can sort these people by Fare in descending order.**

In [6]:
data[(data['Embarked'] == 'C') & 
     (data['Fare'] > 200)].sort_values(by='Fare',
                               ascending=False).head()

Survived  Pclass                                   Name     Sex  \
PassengerId                                                                    
259                 1       1                       Ward, Miss. Anna  female   
680                 1       1     Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1                 Lesurer, Mr. Gustave J    male   
312                 1       1             Ryerson, Miss. Emily Borie  female   
743                 1       1  Ryerson, Miss. Susan Parker "Suzette"  female   

              Age  SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                                  
259          35.0      0      0  PC 17755  512.33              NaN        C  
680          36.0      0      1  PC 17755  512.33      B51 B53 B55        C  
738          35.0      0      0  PC 17755  512.33             B101        C  
312          18.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C  
743          21.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C

**Let's create a new feature.**

In [11]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

In [13]:
age_categories = [age_category(age) for age in data.Age]
data['Age_category'] = age_categories
print('ok')

ok


**Another way is to do it with `apply`.**

In [14]:
data['Age_category'] = data['Age'].apply(age_category)

**1. How many men/women were there onboard?**
- 412 men and 479 women
- 314 men и 577 women
- 479 men и 412 women
- 577 men и 314 women

In [35]:
# You code here
#data['Sex'].count
data['Sex'].value_counts()


male      577
female    314
Name: Sex, dtype: int64

In [37]:
data.head(80)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   
11                  1       3   
12                  1       1   
13                  0       3   
14                  0       3   
15                  0       3   
16                  1       2   
17                  0       3   
18                  1       2   
19                  0       3   
20                  1       3   
21                  0       2   
22                  1       2   
23                  1       3   
24                  1       1   
25                  0       3   
26                  1       3   
27                  0       3   
28                  0       1   
29                  1       3   
30                  0       3   
...               ...     ...   
51                  0       3   
52                  0       3   
53                  1       1   
54                  1       2   
55                  0       1   
56                  1       1   
57                  1       2   
58                  0       3   
59                  1       2   
60                  0       3   
61                  0       3   
62                  1       1   
63                  0       1   
64                  0       3   
65                  0       1   
66                  1       3   
67                  1       2   
68                  0       3   
69                  1       3   
70                  0       3   
71                  0       2   
72                  0       3   
73                  0       2   
74                  0       3   
75                  1       3   
76                  0       3   
77                  0       3   
78                  0       3   
79                  1       2   
80                  1       3   

                                                          Name     Sex    Age  \
PassengerId                                                                     
1                                      Braund, Mr. Owen Harris    male  22.00   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.00   
3                                       Heikkinen, Miss. Laina  female  26.00   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.00   
5                                     Allen, Mr. William Henry    male  35.00   
6                                             Moran, Mr. James    male    NaN   
7                                      McCarthy, Mr. Timothy J    male  54.00   
8                               Palsson, Master. Gosta Leonard    male   2.00   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.00   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.00   
11                             Sandstrom, Miss. Marguerite Rut  female   4.00   
12                                    Bonnell, Miss. Elizabeth  female  58.00   
13                              Saundercock, Mr. William Henry    male  20.00   
14                                 Andersson, Mr. Anders Johan    male  39.00   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.00   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.00   
17                                        Rice, Master. Eugene    male   2.00   
18                                Williams, Mr. Charles Eugene    male    NaN   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.00   
20                                     Masselmani, Mrs. Fatima  female    NaN   
21                                        Fynney, Mr. Joseph J    male  35.00   
22                                       Beesley, Mr. Lawrence    male

**2. Print the distribution of the `Pclass` feature. Then the same, but for men and women separately. How many men from second class were there onboard?**
- 104
- 108
- 112
- 125

In [106]:
# You code here
data[(data['Pclass']==2)& (data['Sex']=='male')].count()

Survived        108
Pclass          108
Name            108
Sex             108
Age              99
SibSp           108
Parch           108
Ticket          108
Fare            108
Cabin             6
Embarked        108
Age_category     99
dtype: int64

**3. What are median and standard deviation of `Fare`?. Round to two decimals.**
- median is  14.45, standard deviation is 49.69
- median is 15.1, standard deviation is 12.15
- median is 13.15, standard deviation is 35.3
- median is  17.43, standard deviation is 39.1

In [52]:
# You code here
print(data['Fare'].median())
print(data['Fare'].std())
data.head(80)

14.4542
49.693428597180905


Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   
11                  1       3   
12                  1       1   
13                  0       3   
14                  0       3   
15                  0       3   
16                  1       2   
17                  0       3   
18                  1       2   
19                  0       3   
20                  1       3   
21                  0       2   
22                  1       2   
23                  1       3   
24                  1       1   
25                  0       3   
26                  1       3   
27                  0       3   
28                  0       1   
29                  1       3   
30                  0       3   
...               ...     ...   
51                  0       3   
52                  0       3   
53                  1       1   
54                  1       2   
55                  0       1   
56                  1       1   
57                  1       2   
58                  0       3   
59                  1       2   
60                  0       3   
61                  0       3   
62                  1       1   
63                  0       1   
64                  0       3   
65                  0       1   
66                  1       3   
67                  1       2   
68                  0       3   
69                  1       3   
70                  0       3   
71                  0       2   
72                  0       3   
73                  0       2   
74                  0       3   
75                  1       3   
76                  0       3   
77                  0       3   
78                  0       3   
79                  1       2   
80                  1       3   

                                                          Name     Sex    Age  \
PassengerId                                                                     
1                                      Braund, Mr. Owen Harris    male  22.00   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.00   
3                                       Heikkinen, Miss. Laina  female  26.00   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.00   
5                                     Allen, Mr. William Henry    male  35.00   
6                                             Moran, Mr. James    male    NaN   
7                                      McCarthy, Mr. Timothy J    male  54.00   
8                               Palsson, Master. Gosta Leonard    male   2.00   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.00   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.00   
11                             Sandstrom, Miss. Marguerite Rut  female   4.00   
12                                    Bonnell, Miss. Elizabeth  female  58.00   
13                              Saundercock, Mr. William Henry    male  20.00   
14                                 Andersson, Mr. Anders Johan    male  39.00   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.00   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.00   
17                                        Rice, Master. Eugene    male   2.00   
18                                Williams, Mr. Charles Eugene    male    NaN   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.00   
20                                     Masselmani, Mrs. Fatima  female    NaN   
21                                        Fynney, Mr. Joseph J    male  35.00   
22                                       Beesley, Mr. Lawrence    male

**4. Is that true that the mean age of survived people is higher than that of passengers who eventually died?**
- Yes
- No


In [151]:
ref=data.dropna()
#data[(data['Survived']==0)]
#print(ref['Age'].mean())
print(ref[ref['Survived']==1]['Age'].mean())
print(ref[ref['Survived']==0]['Age'].mean())

ref.groupby('Survived')['Age'].mean()
#print(data[(data['Survived']==1)].mean())

32.905853658536586
41.35


Survived
0    41.35
1    32.91
Name: Age, dtype: float64

**5. Is that true that passengers younger than 30 y.o. survived more frequently than those older than 60 y.o.? What are shares of survived people among young and old people?**
- 22.7% among young and 40.6% among old
- 40.6% among young and 22.7% among old
- 35.3% among young and 27.4% among old
- 27.4% among young and  35.3% among old

In [177]:
#data[(data['Pclass']==2)& (data['Sex']=='male')].count()
young=data.loc[data['Age'] < 30, 'Survived']
old=data.loc[data['Age'] > 60, 'Survived']


print(round(100*young).mean())
print(round(100*old).mean())


40.625
22.727272727272727


**6. Is that true that women survived more frequently than men? What are shares of survived people among men and women?**
- 30.2% among men and 46.2% among women
- 35.7% among men and 74.2% among women
- 21.1% among men and 46.2% among women
- 18.9% among men and 74.2% among women

In [191]:
# You code here
men=data[data['Sex']=='male']['Survived'].mean()
women=data[data['Sex']=='female']['Survived'].mean()
print(men*100)
print(women*100)

18.890814558058924
74.20382165605095


**7. What's the most popular first name among male passengers?**
- Charles
- Thomas
- William
- John

In [210]:
# You code here
first_names = data.loc[data['Sex'] == 'male', 'Name'].apply(lambda 
                                                            full_name: 
                full_name.split(',')[1].split()[1])
first_names.value_counts().head()

William    35
John       25
George     14
Charles    13
Thomas     13
Name: Name, dtype: int64

**8. How is average age for men/women dependent on `Pclass`? Choose all correct statements:**
- On average, men of 1 class are older than 40
- On average, women of 1 class are older than 40
- Men of all classes are on average older than women of the same class
- On average, passengers ofthe first class are older than those of the 2nd class who are older than passengers of the 3rd class

In [212]:
# You code here

pd.crosstab(data['Pclass'], data['Sex'],values=data['Age'], aggfunc=np.mean)

Sex     female   male
Pclass               
1        34.61  41.28
2        28.72  30.74
3        21.75  26.51

## Useful resources
* The same notebook as an interactive web-based [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-1-practice-analyzing-titanic-passengers) with a [solution](https://www.kaggle.com/kashnitsky/topic-1-practice-solution)
* Topic 1 "Exploratory Data Analysis with Pandas" as a [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-1-exploratory-data-analysis-with-pandas)
* Main course [site](https://mlcourse.ai), [course repo](https://github.com/Yorko/mlcourse.ai), and YouTube [channel](https://www.youtube.com/watch?v=QKTuw4PNOsU&list=PLVlY_7IJCMJeRfZ68eVfEcu-UcN9BbwiX)